lapis/src/main/kotlin/org/genspectrum/lapis/model/AdvancedQueryCustomListener.kt

```diff
@@ -0,0 +1,516 @@
+
+        var name = metadataName
+        if (metadataName.endsWith(".regex")) {
+            name = metadataName.substringBeforeLast(".regex")
+        }
+
+        val field: DatabaseMetadata =
+            metadataFieldsByName[name.lowercase(Locale.US)]
+                ?: throw BadRequestException("Metadata field $metadataName does not exist", null)
+        when (field.type) {
+            MetadataType.STRING -> {
+                if (field.generateLineageIndex) {
+                    expressionStack.addLast(mapToLineageFilter(field.name, metadataValue))
+                } else {
+                    if (field.lapisAllowsRegexSearch and metadataName.endsWith(".regex")) {
+                        expressionStack.addLast(StringSearch(field.name, metadataValue))
+                    } else if (metadataName.endsWith(".regex")) {
+                        throw BadRequestException(
+                            "Metadata field `${field.name}` does not support regex search.",
+                            null,
+                        )
+                    } else {
+                        expressionStack.addLast(StringEquals(field.name, metadataValue))
+                    }
+                }
+            }
+
+            MetadataType.BOOLEAN -> {
+                try {
+                    expressionStack.addLast(BooleanEquals(field.name, metadataValue.toBoolean()))
+                } catch (e: IllegalArgumentException) {
+                    throw BadRequestException("'$metadataValue' is not a valid boolean", e)
+                }
+            }
+
+            MetadataType.DATE -> {
+                try {
+                    val date = LocalDate.parse(metadataValue)
+                    expressionStack.addLast(DateBetween(field.name, to = date, from = date))
+                } catch (exception: DateTimeParseException) {
+                    throw BadRequestException("'$metadataValue' is not a valid date: ${exception.message}", exception)
+                }
+            }
+
+            MetadataType.FLOAT -> {
+                try {
+                    expressionStack.addLast(FloatEquals(field.name, metadataValue.toDouble()))
+                } catch (e: NumberFormatException) {
+                    throw BadRequestException("'$metadataValue' is not a valid float", e)
+                }
+            }
+
+            MetadataType.INT -> {
+                try {
+                    expressionStack.addLast(IntEquals(field.name, metadataValue.toInt()))
+                } catch (e: NumberFormatException) {
+                    throw BadRequestException("'$metadataValue' is not a valid integer", e)
+                }
+            }
+        }
+    }
+
+    override fun enterIsNullQuery(ctx: AdvancedQueryParser.IsNullQueryContext) {
+        val metadataName = ctx.geneOrName().text
+
+        var name = metadataName
+        if (metadataName.endsWith(".regex")) {
+            name = metadataName.substringBeforeLast(".regex")
+        }
```

Comment:
Do we really need to consider this case here? IMO we could get rid of this and simply let the error message `Metadata field author.regex does not exist` be enough.

(I think of the .regex case not as a "metadata field", but as a separate case that happens to use the same parsing logic.)

Reply:
I think this gives a clearer error if someone tries to search IsNull(metadata.regex) for example - I would prefere to keep it

Reply:
IMO we don't need that. It's essentially the same as a `dateFrom`: It's not a database field. A user should not get the impression that you can use it as one. (That's also why I don't like our current regex solution that much).

If we do want to keep it here, let's rather directly throw an exception here.